File: tools/mlvc/pmf.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (c) 2026 梦归云帆
"""Microsoft MLVC PMF JSON ↔ rkvc ``PMF1`` 二进制表。

二进制布局与 ``lib/node_mlvc.c`` ``load_pmf()`` 一致：

    magic "PMF1"
    uint32 nL, nO, nT
    int32  lengths[nL], offsets[nO], table[nT]
    tag=1  gaussian: double scale_min, scale_max; uint32 scale_levels, index_space
    tag=2  bitest:   uint32 qp_num, channels
"""

from __future__ import annotations

import json
import struct
from pathlib import Path
from typing import Any, Mapping, Sequence, Union

PathLike = Union[str, Path]

PMF_MAGIC = b"PMF1"
TAG_GAUSSIAN = 1
TAG_BITEST = 2

_I32 = struct.Struct("<i")
_U32 = struct.Struct("<I")
_F64 = struct.Struct("<d")


class PmfError(ValueError):
    """PMF 文件格式或字段错误。"""


def _as_int_list(name: str, values: Sequence[Any]) -> list[int]:
    out: list[int] = []
    for i, raw in enumerate(values):
        try:
            out.append(int(raw))
        except (TypeError, ValueError) as exc:
            raise PmfError(f"{name}[{i}] 不是整数: {raw!r}") from exc
    return out


def _require(data: Mapping[str, Any], *keys: str) -> None:
    missing = [k for k in keys if k not in data]
    if missing:
        raise PmfError(f"JSON 缺少字段: {', '.join(missing)}")


def load_json(path: PathLike) -> dict[str, Any]:
    src = Path(path)
    with src.open(encoding="utf-8") as fh:
        data = json.load(fh)
    if not isinstance(data, dict):
        raise PmfError(f"{src}: 顶层必须是 JSON object")
    return data


def detect_kind(data: Mapping[str, Any]) -> str:
    if "scale_min" in data or "scale_levels" in data or "index_space" in data:
        return "gaussian"
    if "qp_num" in data or "channels" in data:
        return "bitest"
    raise PmfError("无法判断 PMF 类型：需要 gaussian(scale_min/scale_levels) 或 bitest(qp_num/channels)")


def json_to_pmf1(data: Mapping[str, Any], *, kind: str | None = None) -> bytes:
    kind = kind or detect_kind(data)
    _require(data, "pmf_lengths", "pmf_offsets", "pmf_table")
    lengths = _as_int_list("pmf_lengths", data["pmf_lengths"])
    offsets = _as_int_list("pmf_offsets", data["pmf_offsets"])
    table = _as_int_list("pmf_table", data["pmf_table"])

    chunks = [PMF_MAGIC]
    chunks.append(_U32.pack(len(lengths)))
    chunks.append(_U32.pack(len(offsets)))
    chunks.append(_U32.pack(len(table)))
    chunks.append(b"".join(_I32.pack(v) for v in lengths))
    chunks.append(b"".join(_I32.pack(v) for v in offsets))
    chunks.append(b"".join(_I32.pack(v) for v in table))

    if kind == "gaussian":
        _require(data, "scale_min", "scale_max", "scale_levels")
        index_space = data.get("index_space", False)
        chunks.append(_U32.pack(TAG_GAUSSIAN))
        chunks.append(_F64.pack(float(data["scale_min"])))
        chunks.append(_F64.pack(float(data["scale_max"])))
        chunks.append(_U32.pack(int(data["scale_levels"])))
        chunks.append(_U32.pack(1 if index_space else 0))
    elif kind == "bitest":
        _require(data, "qp_num", "channels")
        qp_num = int(data["qp_num"])
        channels = int(data["channels"])
        if qp_num * channels != len(lengths):
            raise PmfError(
                f"bitest 尺寸不一致: qp_num*{channels}={qp_num * channels} != len(pmf_lengths)={len(lengths)}"
            )
        chunks.append(_U32.pack(TAG_BITEST))
        chunks.append(_U32.pack(qp_num))
        chunks.append(_U32.pack(channels))
    else:
        raise PmfError(f"未知 PMF 类型: {kind}")

    return b"".join(chunks)


def pmf1_to_dict(blob: bytes) -> dict[str, Any]:
    if len(blob) < 16 or blob[:4] != PMF_MAGIC:
        raise PmfError("不是 PMF1 文件")
    nL, nO, nT = struct.unpack_from("<III", blob, 4)
    off = 16
    need = off + (nL + nO + nT) * 4 + 4
    if len(blob) < need:
        raise PmfError("PMF1 文件截断")

    def _take_i32(count: int) -> list[int]:
        nonlocal off
        vals = list(struct.unpack_from("<" + "i" * count, blob, off))
        off += count * 4
        return vals

    lengths = _take_i32(nL)
    offsets = _take_i32(nO)
    table = _take_i32(nT)
    tag = struct.unpack_from("<I", blob, off)[0]
    off += 4
    base = {
        "pmf_lengths": lengths,
        "pmf_offsets": offsets,
        "pmf_table": table,
    }
    if tag == TAG_GAUSSIAN:
        if len(blob) < off + 24:
            raise PmfError("gaussian 尾部截断")
        scale_min, scale_max = struct.unpack_from("<dd", blob, off)
        scale_levels, index_space = struct.unpack_from("<II", blob, off + 16)
        base.update(
            {
                "kind": "gaussian",
                "scale_min": scale_min,
                "scale_max": scale_max,
                "scale_levels": scale_levels,
                "index_space": bool(index_space),
            }
        )
        return base
    if tag == TAG_BITEST:
        if len(blob) < off + 8:
            raise PmfError("bitest 尾部截断")
        qp_num, channels = struct.unpack_from("<II", blob, off)
        base.update({"kind": "bitest", "qp_num": qp_num, "channels": channels})
        return base
    raise PmfError(f"未知 PMF1 tag={tag}")


def convert_json_file(src: PathLike, dst: PathLike, *, kind: str | None = None) -> dict[str, Any]:
    data = load_json(src)
    resolved = kind or detect_kind(data)
    blob = json_to_pmf1(data, kind=resolved)
    out = Path(dst)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_bytes(blob)
    meta = pmf1_to_dict(blob)
    meta["source"] = str(Path(src))
    meta["output"] = str(out)
    meta["bytes"] = len(blob)
    return meta

File: lib/pmf.c
/* SPDX-License-Identifier: AGPL-3.0-or-later */
/* Copyright (c) 2026 梦归云帆 */
/*
 * MLVC PMF1 二进制表加载（布局与 tools/mlvc/pmf.py 一致）：
 *
 *   magic "PMF1"
 *   uint32 nL, nO, nT
 *   int32  lengths[nL], offsets[nO], table[nT]
 *   tag=1  gaussian: double scale_min, scale_max; uint32 scale_levels, index_space
 *   tag=2  bitest:   uint32 qp_num, channels
 *
 * 从 lib/node_mlvc.c 抽出，使 Python 工具产物（tools/mlvc/pmf.py）与 C
 * 消费方可由同一解析代码在契约测试中互相校验。
 */

#include "pmf.h"
#include "internal.h"

#include <stdio.h>
#include <string.h>

/* lengths/offsets 上界 1M 项；table 上界 16M 项（约 64MB），挡住计数回绕后的 fread 堆溢出。 */
#define RKVC_PMF_MAX_LEN  (1u << 20)
#define RKVC_PMF_MAX_TAB  (16u << 20)

void rkvc_pmf_free(rkvc_pmf *p)
{
    if (!p)
        return;
    rkvc_free(p->lengths);
    rkvc_free(p->offsets);
    rkvc_free(p->table);
    memset(p, 0, sizeof(*p));
}

rkvc_err rkvc_pmf_load(rkvc_pmf *p, const char *path)
{
    FILE *f = NULL;
    rkvc_err err = RKVC_ERR_FORMAT;

    memset(p, 0, sizeof(*p));
    f = fopen(path, "rb");
    if (!f)
        return RKVC_ERR_IO;

    char magic[4];
    if (fread(magic, 1, 4, f) != 4 || memcmp(magic, "PMF1", 4) != 0)
        goto fail;

    uint32_t nL, nO, nT;
    if (fread(&nL, 4, 1, f) != 1 || fread(&nO, 4, 1, f) != 1 ||
        fread(&nT, 4, 1, f) != 1)
        goto fail;
    if (nL == 0 || nL > RKVC_PMF_MAX_LEN ||
        nO == 0 || nO > RKVC_PMF_MAX_LEN ||
        nT == 0 || nT > RKVC_PMF_MAX_TAB)
        goto fail;

    p->num_lengths = nL;
    p->num_offsets = nO;
    p->num_table = nT;
    p->lengths = rkvc_malloc((size_t)nL * 4u);
    p->offsets = rkvc_malloc((size_t)nO * 4u);
    p->table   = rkvc_malloc((size_t)nT * 4u);
    if (!p->lengths || !p->offsets || !p->table) {
        err = RKVC_ERR_NOMEM;
        goto fail;
    }
    if (fread(p->lengths, 4, nL, f) != nL ||
        fread(p->offsets, 4, nO, f) != nO ||
        fread(p->table, 4, nT, f) != nT)
        goto fail;

    uint32_t tag = 0;
    if (fread(&tag, 4, 1, f) != 1)
        goto fail;
    if (tag == 1) {
        if (fread(&p->scale_min, 8, 1, f) != 1 ||
            fread(&p->scale_max, 8, 1, f) != 1 ||
            fread(&p->scale_levels, 4, 1, f) != 1 ||
            fread(&p->index_space, 4, 1, f) != 1)
            goto fail;
    } else if (tag == 2) {
        if (fread(&p->qp_num, 4, 1, f) != 1 ||
            fread(&p->channels, 4, 1, f) != 1)
            goto fail;
    } else {
        goto fail;
    }
    fclose(f);
    return RKVC_OK;

fail:
    fclose(f);
    rkvc_pmf_free(p);
    return err;
}
